### PR TITLE
Add API Gateway, Queue and Lambda support

### PR DESCRIPTION
**Pull request by coverbeck**
SEAB-789

Separated more persistent resources into two additional templates,
log-group.yml and gateway-and-queue.yml.

Started updating README with more instructions, but more to go.

Got rid of the word 'port-' from target group names, to allow for longer
stack names (the maximum length of a target group name is 32 chars).

Split out two tickets, dockstore/dockstore#3326, to add logging to
API Gateway, and dockstore/dockstore#3324, to auto upload lambdas.

For now, manually created a ZIP deployment of the lambda and put
it on s3.

Use parameter store settings for everything in compose_setup; was
adding them piecemeal before. Not all the values are correctly set
in parameter store itself, but now it's just a question of
filling them in.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,76 @@
...
+One is for for a legacy environment with resources created not using CloudFormation.
+
+### Templates for a complete Dockstore deploy
+
+1. log-group.yml - This should only be run once; the created log group should persist indefinitely.
+1. gateway-and-queue.yml - This is for deploying the API Gateway and SQS queues that process GitHub app requests. It is not 
+expected that this will change often, unlike frequent deploys of Dockstore. In addition, we want the Gateway and queue
+to persist across Dockstore version deploys, e.g., a message that came in during one Dockstore version's deploy and for some
+reason didn't get processed should still be in the queue for the next Dockstore version.
+1. dockstore.yml - This is for deploying the rest of the AWS resources needed to run Dockstorel. The expected flow is that you will use this template to create a deployment of a particular
```
Comment (denis-yuen) [inline README.md]: Dockstorel typo

File: README.md [hunk 2]
```diff
@@ -1,2 +1,76 @@
...
+
+Figure out what version of the UI, web service, compose_setup, and Galaxy plugin you want to deploy. For a non-release
+instance, you will typically want to take the defaults, **except for the UI**. Because the UI is frequently built, sometimes
+more than once daily, and the build has the Git commit SHA in it, the default value in dockstore.yml is most certainly not the
+most recent. If you want the latest build off develop that has been deployed to CloudFront, then run the
+`latest_ui_cloudfront.sh` file that is part of this repo (node.js installation required).
+
+## Deploy
+
+Currently using the following convention for the stack names:
```
Comment (denis-yuen) [inline README.md]: Not clear to me, maybe its a parameter store. But how do we determine that "this" gateway and queue is for "this" dockstore if there are multiple ones in an account. Probably explained later, but that's the question I had at this juncture

File: dockstore.yml [hunk 3]
```diff
@@ -1,4 +1,4 @@
-# A template to a create a Dockstore EC2 instance
+# A template to a create AWS resources for running Dockstore
```
Comment (denis-yuen) [inline dockstore.yml]: "A template to a create" was like this before, but

File: README.md [hunk 4]
```diff
@@ -1,2 +1,76 @@
 # dockstore-deploy
-A sandbox repo for deploying Dockstore via Cloudformation
+A repo for deploying Dockstore via Cloudformation
```
Comment (garyluu) [inline README.md]: Cloud​Formation

File: README.md [hunk 5]
```diff
@@ -1,2 +1,76 @@
...
+  * gateway-and-queue.yml: gateway-<MMDD>, e.g., gateway-0317
+  * dockstore.yml: dockstore-<MMDD>, e.g, dockstore-0317
+
+Ideally would use long-winded descriptive names, but the stack name gets
+made part of some AWS resource names such as target groups, which can only be 32 characters long.
+
+### Order to run the templates 
+
+1. log-group.yml - but you will only do this once per domain, i.e., once you have run in it for dev.dockstore.net, you never need to run it again.
+1. gateway-and-queue.yml - this will typically be run unfrequently, although probably not as infrequently as the previous item. Probably only if the lambda changes, which is rare.
```
Comment (garyluu) [inline README.md]: unfrequently, infrequently

File: dockstore.yml [hunk 6]
```diff
@@ -79,17 +72,81 @@ Parameters:
...
+## Application level parameters with values in Parameter Store
+  BdCatalystSevenBridgesImportUrl:
+    Description: The import url for the Seven Bridges instance of BD Catalyst
+    Type: AWS::SSM::Parameter::Value<String>
+    Default: /DeploymentConfig/dev/BdCatalystSevenBridgesImportUrl
+  BdCatalystTerraImportUrl:
+    Description: The import url for the Terra instance of BD Catalyst
+    Type: AWS::SSM::Parameter::Value<String>
+    Default: /DeploymentConfig/dev/BdCatalystTerraImportUrl
+  BitBucketClientId:
```
Comment (garyluu) [inline dockstore.yml]: BitbucketClientId

File: README.md (hunk 2 above)
Comment (coverbeck) [reply README.md]: Yes, it's determined by parameter store. dockstore.yml has:

```
 ...
 LambdaFunctionEventSourceMapping:
    Type: AWS::Lambda::EventSourceMapping
    Properties:
      ...
      EventSourceArn: !Join
        - ''
        - - 'arn:aws:sqs:'
          - !Ref 'AWS::Region'
          - ':'
          - !Ref 'AWS::AccountId'
          - ':'
          - !Ref 'WebhookQueueName'
      FunctionName: !GetAtt GitHubQueueLambda.Arn 
```

The value of `WebhookQueueName` comes from parameter store; which gateway-and-queue.yml used to create a queue with that name, and the lambda uses it as a source.

File: dockstore.yml (hunk 3 above)
Comment (coverbeck) [reply dockstore.yml]: Looks like you didn't finish your thought?

Comment (denis-yuen) [reply dockstore.yml]: No, that is it. Should probably be "A template to create"

**Pull request merged**